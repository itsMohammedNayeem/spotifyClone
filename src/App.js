import React, {
  useState,
  Suspense,
  createContext,
  useEffect,
  useReducer,
  useRef
} from 'react'
import { initialState, reducer } from './reducers'
import './App.css'
import { login, logout, selectUser } from './features/userSlice'
import Info from './Info'
import Home from './Home'
import Header from './Header'
import Login from './Login'
import { useDispatch, useSelector } from 'react-redux'
import { auth } from './firebase'
import Widgets from './Widgets'
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'
import NotFoundPage from './NotFoundPage'
import api from './api/songs'
import { v4 as uuidv4 } from 'uuid'
import AddSong from './AddSong'
import EditSong from './EditSong'
import SongDetail from './SongDetail'
import Trend from './Trend'
import Sidebar from './Sidebar'
import { Global, css, jsx } from '@emotion/core'
import MusicPlayer from './musicplayer'

const SongList = React.lazy(() => import('./SongList'))

export const StoreContext = createContext(null)

function App() {
  const user = useSelector(selectUser)
  const dispatch1 = useDispatch()
  const [state, dispatch] = useReducer(reducer, initialState)

  const [songs, setSongs] = useState([])
  const [searchTerm, setSearchTerm] = useState('')
  const [searchResults, setSearchResults] = useState([])
  const [view, setView] = useState(true)

  const retrieveSongs = async () => {
    const response = await api.get('/songs')
    setSongs(response.data)
    return response.data
  }

  const addSongHandler = async (song) => {
    const request = {
      id: uuidv4(),
      ...song,
      count: 1
    }

    const response = await api.post('/songs', request)
    setSongs([...songs, response.data])
  }

  const updateSongHandler = async (song) => {
    const response = await api.put(`/songs/${song.id}`, song)
    const { id } = response.data

    setSongs(
      songs.map((song) => {
        return song.id === id ? { ...response.data } : song
      })
    )
  }

  const removeSongHandler = async (id) => {
    await api.delete(`/songs/${id}`)
    const newSongList = songs.filter((song) => {
      return song.id !== id
    })
    setSongs(newSongList)
  }

  const searchHandler = (searchTerm) => {
    setSearchTerm(searchTerm)
    if (searchTerm !== '') {
      const newSongList = songs.filter((contact) => {
        return Object.values(contact)
          .join(' ')
          .toLowerCase()
          .includes(searchTerm.toLowerCase())
      })
      setSearchResults(newSongList)
    } else {
      setSearchResults(songs)
    }
  }

  useEffect(() => {
    const getAllSongs = async () => {
      const allSongs = await retrieveSongs()
      if (allSongs) setSongs(allSongs)
    }

    getAllSongs()
  }, [])

  // useEffect(() => {
  //   auth.onAuthStateChanged((userAuth) => {
  //     console.log('auth changed', userAuth)
  //   })
  // }, [])

  useEffect(() => {
    auth.onAuthStateChanged((userAuth) => {
      if (userAuth) {
        dispatch1(
          login({
            email: userAuth.email,
            uid: userAuth.uid,
            displayName: userAuth.displayName,
            photoUrl: userAuth.photoURL
          })
        )
      } else {
        dispatch(logout())
      }
    })
  }, [dispatch1])

  const showView = () => {
    if (window.innerWidth <= 960) {
      setView(false)
    } else {
      setView(true)
    }
  }

  useEffect(() => {
    showView()
    window.addEventListener('resize', showView)
  }, [])

  return (
    <>
      <StoreContext.Provider value={{ state, dispatch }}>
        <Router>
          {!user ? (
            <div className="app">
              <Login />
            </div>
          ) : (
            <div className="app">
              <Header />
              <div className="app__body">
                {view ? <Sidebar /> : null}
                <Suspense
                  fallback={
                    <div className="home">
                      <div className="home__inputContainer">
                        <img
                          src="https://cdn.dribbble.com/users/189524/screenshots/1887541/foxfunwalk-800x600_v2.gif"
                          alt="Loading..."
                        />
                      </div>
                    </div>
                  }
                >
                  <Switch>
                    <Route exact path="/" component={Home} />
                    <Route exact path="/info" component={Info} />
                    <Route
                      exact
                      path="/trend"
                      render={() => <Trend songs={songs} />}
                    />
                    <Route
                      path="/list"
                      exact
                      render={(props) => (
                        <SongList
                          {...props}
                          songs={searchTerm.length < 1 ? songs : searchResults}
                          getSongId={removeSongHandler}
                          term={searchTerm}
                          searchKeyword={searchHandler}
                          updateSongHandler={updateSongHandler}
                          retrieveSongs={retrieveSongs}
                        />
                      )}
                    />
                    <Route exact path="/playlist" component={MusicPlayer} />

                    <Route
                      path="/edit"
                      exact
                      render={(props) => (
                        <EditSong
                          {...props}
                          updateSongHandler={updateSongHandler}
                        />
                      )}
                    />
                    <Route path="/song/:id" component={SongDetail} />
                    <Route
                      path="/add"
                      render={(props) => (
                        <AddSong {...props} addSongHandler={addSongHandler} />
                      )}
                    />
                    <Route path="*" component={NotFoundPage} />
                  </Switch>
                </Suspense>
                {view ? <Widgets /> : null}
              </div>
            </div>
          )}
        </Router>
      </StoreContext.Provider>
    </>
  )
}

export default App
