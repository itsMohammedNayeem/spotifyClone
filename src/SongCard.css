.songcard {
  background-color: white;
  padding: 20px;
  padding-bottom: 20px;
  border-radius: 10px;
  margin-bottom: 20px;
}

.songcard_detail {
  display: flex;
  flex-wrap: inherit;
  justify-content: space-between;
  margin-bottom: 20px;
  /* align-items: center; */
  /* font-size: 1.25vw; */
}

.songcard_buttons {
  display: flex;
  flex-wrap: unset;
  justify-content: space-between;
  align-items: center;
}

.songcard__comment {
  padding: 10px;
  display: flex;
  flex-wrap: initial;
  /* align-items: center; */
  border-radius: 25px;
  height: 15px;
  width: 75%;
  color: gray;
  background-color: #eef3f8;
}

.songcard__comment > form > input {
  /* display: flex;
  flex-wrap: unset; */
  outline: none;
  border: none;
  background: none;
  /* font-weight: 2vw; */
  margin-left: 5px;
}

.songcard__comment > form > button {
  outline: none;
  border: none;
  background: none;
}
