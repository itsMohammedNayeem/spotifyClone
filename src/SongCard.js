import React, { useState, useEffect } from 'react'
import { withRouter, Link } from 'react-router-dom'
import EditIcon from '@material-ui/icons/Edit'
import DeleteIcon from '@material-ui/icons/Delete'
import AlbumIcon from '@material-ui/icons/Album'
import MusicNoteIcon from '@material-ui/icons/MusicNote'
import HistoryIcon from '@material-ui/icons/History'
import RecordVoiceOverIcon from '@material-ui/icons/RecordVoiceOver'
import './SongCard.css'
import AddCommentIcon from '@material-ui/icons/AddComment'
import Post from './Post'
import { db } from './firebase'
import firebase from 'firebase'
import { useSelector } from 'react-redux'
import { selectUser } from './features/userSlice'
import { IconButton } from '@material-ui/core'
import api from './api/songs'

const SongCard = (props) => {
  const { id, title, movie, duration, artist, count } = props.song
  const user = useSelector(selectUser)
  const [posts, setPosts] = useState([])
  const [input, setInput] = useState('')

  useEffect(() => {
    db.collection('posts')
      .orderBy('timestamp', 'desc')
      .onSnapshot((snapshot) =>
        setPosts(
          snapshot.docs.map((doc) => ({
            id: doc.id,
            data: doc.data()
          }))
        )
      )
    return () => {
      setPosts([])
    }
  }, [])

  const updateSongHandler = async (songs) => {
    await api.put(`/songs/${songs.id}`, songs)
    props.retrieveSongs()
    // const { id, title, movie, duration, artist, count } = response.data;
    // setSongs(
    //   songs.map((song) => {
    //     return song.id === id ? { ...response.data } : song;
    //   })
    // );
  }

  // const sendPost = (e) => {
  //   e.preventDefault()

  //   db.collection('posts').add({
  //     name: user.displayName,
  //     description: user.email,
  //     message: input,
  //     photoUrl: user.photoUrl || '',
  //     timestamp: firebase.firestore.FieldValue.serverTimestamp()
  //   })

  //   setInput('')
  // }

  return (
    <>
      <div className="songcard">
        <div className="songcard_detail">
          {props.clickTitle ? (
            <MusicNoteIcon style={{ color: 'E99497' }} />
          ) : null}
          <Link
            to={{
              pathname: `/song/${id}`,
              state: { title, movie, duration, artist, count }
            }}
            style={{ textDecoration: 'none' }}
            onClick={() =>
              updateSongHandler({
                id,
                title,
                movie,
                duration,
                artist,
                count: count + 1
              })
            }
          >
            {title}
          </Link>
          {props.clickMovie ? <AlbumIcon style={{ color: 'F3C583' }} /> : null}
          {props.clickMovie ? movie : null}

          {props.clickDuration ? (
            <HistoryIcon style={{ color: 'E8E46E' }} />
          ) : null}
          {props.clickDuration ? duration : null}

          {props.clickArtist ? (
            <RecordVoiceOverIcon style={{ color: 'B3E283' }} />
          ) : null}
          {props.clickArtist ? artist : null}
        </div>
        <div className="songcard_buttons">
          <Link
            to={{
              pathname: `/edit`,
              state: { id, title, movie, duration, artist, count }
            }}
            style={{ textDecoration: 'none' }}
          >
            <IconButton>
              <EditIcon style={{ color: '89CFF0' }} />
            </IconButton>
          </Link>
          {/* <div className="songcard__comment">
            <AddCommentIcon style={{ color: "gray" }} />
            <form>
              <input
                value={input}
                onChange={(e) => setInput(e.target.value)}
                type="text"
                placeholder="Comment"
              />
              <button onClick={sendPost} type="submit"></button>
            </form>
          </div> */}
          <IconButton
            onClick={() => {
              if (window.confirm('Remove from the playlist?')) {
                props.clickTitleHandler(id)
              }
            }}
          >
            <DeleteIcon style={{ color: 'FF7779' }} />
          </IconButton>
        </div>
        {posts.map(({ id, data: { name, description, message, photoUrl } }) => (
          <Post
            key={id}
            name={name}
            description={description}
            message={message}
            photoUrl={photoUrl}
          />
        ))}
      </div>
    </>
  )
}

export default withRouter(SongCard)
