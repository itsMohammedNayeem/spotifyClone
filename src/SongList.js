import React, { useRef, useState } from 'react'
import { Link } from 'react-router-dom'
import SongCard from './SongCard'
import AddIcon from '@material-ui/icons/Add'
import SearchIcon from '@material-ui/icons/Search'
import './SongList.css'
import AlbumIcon from '@material-ui/icons/Album'
import MusicNoteIcon from '@material-ui/icons/MusicNote'
import HistoryIcon from '@material-ui/icons/History'
import RecordVoiceOverIcon from '@material-ui/icons/RecordVoiceOver'
import Checkbox from '@material-ui/core/Checkbox'

const SongList = (props) => {
  const [clickTitle, setClickTitle] = useState(true)
  const [clickMovie, setClickMovie] = useState(true)
  const [clickDuration, setClickDuration] = useState(true)
  const [clickArtist, setClickArtist] = useState(true)
  const inputEl = useRef('')
  const retrieveSongs = props.retrieveSongs

  const handleClickTitle = () => setClickTitle(!clickTitle)
  const handleClickMovie = () => setClickMovie(!clickMovie)
  const handleClickDuration = () => setClickDuration(!clickDuration)
  const handleClickArtist = () => setClickArtist(!clickArtist)

  const deleteSongHandler = (id) => {
    props.getSongId(id)
  }

  const renderSongList = props.songs.map((song) => {
    return (
      <SongCard
        song={song}
        retrieveSongs={retrieveSongs}
        clickTitleHandler={deleteSongHandler}
        key={song.id}
        clickTitle={clickTitle}
        clickMovie={clickMovie}
        clickDuration={clickDuration}
        clickArtist={clickArtist}
      />
    )
  })

  const getSearchTerm = () => {
    props.searchKeyword(inputEl.current.value)
  }

  return (
    <div className="songlist">
      <div className="songlist__stats">
        <div className="songlist__stat">
          <h2>Your Library</h2>
          <div className="songlist__search">
            <SearchIcon />
            <input
              id="search"
              ref={inputEl}
              type="text"
              placeholder="Artists, songs, or album"
              value={props.term}
              onChange={getSearchTerm}
            />
            <Link to="/add" style={{ textDecoration: 'none' }}>
              <AddIcon style={{ color: 'gray' }} />
            </Link>
          </div>
        </div>
      </div>
      <div className="songlist__hide">
        <h4
          style={{
            paddingLeft: '20px',
            paddingTop: '10px',
            paddingBottom: '10px'
          }}
        >
          Hide options
        </h4>
        <div className="songlist__hidebuttons">
          {/* <MusicNoteIcon style={{ color: 'E99497' }} />
          <Checkbox
            onClick={handleClickTitle}
            defaultChecked
            size="small"
            color="primary"
            inputProps={{ 'aria-label': 'checkbox with small size' }}
          /> */}
          <AlbumIcon style={{ color: 'F3C583' }} />
          <Checkbox
            defaultChecked
            onClick={handleClickMovie}
            size="small"
            color="primary"
            inputProps={{ 'aria-label': 'checkbox with small size' }}
          />
          <HistoryIcon style={{ color: 'E8E46E' }} />
          <Checkbox
            defaultChecked
            onClick={handleClickDuration}
            size="small"
            color="primary"
            inputProps={{ 'aria-label': 'checkbox with small size' }}
          />
          <RecordVoiceOverIcon style={{ color: 'B3E283' }} />
          <Checkbox
            onClick={handleClickArtist}
            defaultChecked
            size="small"
            color="primary"
            inputProps={{ 'aria-label': 'checkbox with small size' }}
          />
        </div>
      </div>
      {renderSongList.length > 0 ? (
        renderSongList
      ) : (
        <>
          <h4
            style={{ color: '#FF7779', padding: '10px' }}
          >{`Couldn't find "${props.term}"`}</h4>
          <p style={{ color: 'gray', paddingLeft: '10px', fontSize: '12px' }}>
            Try the right term.
          </p>
        </>
      )}
    </div>
  )
}

export default SongList
