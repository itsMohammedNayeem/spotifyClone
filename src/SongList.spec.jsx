import React from "react";
import { mount, shallow } from "enzyme";
import SongList from "./SongList";
import SongCard from "./SongCard";
import App from "./App";

describe("SongList", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<div />);
  });

  test("render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("displays div", () => {
    expect(wrapper.find("div").length).toEqual(1);
  });
});

describe("When songs array passed to SongsList is empty", () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      song: [],
    };
    wrapper = shallow(<SongCard {...props} />);
  });

  it("should not crash", () => {
    let li = wrapper.find("song");
    expect(li.length).toEqual(0);
  });
});

describe("When songs array is passed to SongsList", () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      songs: [
        {
          id: 1,
          title: "Tumhi Dekho Naa",
          movie: "Kabhi Alvida Na Kehna",
          duration: "5:35",
          artist: "Sonu Nigam",
          count: 14,
        },
        {
          id: "2",
          title: "Mitwa",
          movie: "Kabhi Alvida Na Kehna",
          duration: "2:30",
          artist: "Shafqat Ali",
          count: 5,
        },
      ],
    };
    wrapper = shallow(<SongCard {...props} />);
  });

  it("renders correctly when passed the props", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
