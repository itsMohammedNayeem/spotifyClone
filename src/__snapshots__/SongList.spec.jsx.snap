// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongList render correctly 1`] = `ShallowWrapper {}`;

exports[`When songs array is passed to SongsList renders correctly when passed the props 1`] = `ShallowWrapper {}`;
